kv/kvserver/rangefeed: TestProcessorBackpressure failed

kv/kvserver/rangefeed.TestProcessorBackpressure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18007415?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18007415?buildTab=artifacts#/) on master @ [c14af3c5756e94ca09a0dff809ac20fb97549980](https://github.com/cockroachdb/cockroach/commits/c14af3c5756e94ca09a0dff809ac20fb97549980):

Fatal error:

```
panic: test timed out after 5m0s
running tests:
	TestProcessorBackpressure (4m59s)
	TestProcessorBackpressure/feed_type=scheduled (4m59s)

```
Stack: 

```
goroutine 527 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestProcessorBackpressure
=== RUN   TestProcessorBackpressure/feed_type=scheduled
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestProcessorBackpressure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45112

--- a/pkg/kv/kvserver/rangefeed/buffered_registration.go
+++ b/pkg/kv/kvserver/rangefeed/buffered_registration.go
@@ -8,6 +8,7 @@ package rangefeed
 import (
 	"context"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
@@ -52,9 +53,6 @@ type bufferedRegistration struct {
 		// This will cause the registration to exit with an error once the buffer
 		// has been emptied.
 		overflowed bool
-		// Boolean indicating if all events have been output to stream. Used only
-		// for testing.
-		caughtUp bool
 		// Management of the output loop goroutine, used to ensure proper teardown.
 		outputLoopCancelFn func()
 		disconnected       bool
@@ -65,6 +63,10 @@ type bufferedRegistration struct {
 		// that disconnect is called, it is closed by disconnect.
 		catchUpIter *CatchUpIterator
 	}
+
+	// Number of events that have been written to the buffer but
+	// not sent. Used only for testing.
+	testPendingEventToSend atomic.Int64
 }
 
 var _ registration = &bufferedRegistration{}
@@ -99,7 +101,6 @@ func newBufferedRegistration(
 		blockWhenFull: blockWhenFull,
 	}
 	br.mu.Locker = &syncutil.Mutex{}
-	br.mu.caughtUp = true
 	br.mu.catchUpIter = catchUpIter
 	return br
 }
@@ -124,7 +125,7 @@ func (br *bufferedRegistration) publish(
 	alloc.Use(ctx)
 	select {
 	case br.buf <- e:
-		br.mu.caughtUp = false
+		br.testPendingEventToSend.Add(1)
 	default:
 		// If we're asked to block (in tests), do a blocking send after releasing
 		// the mutex -- otherwise, the output loop won't be able to consume from the
@@ -135,7 +136,7 @@ func (br *bufferedRegistration) publish(
 			select {
 			case br.buf <- e:
 				br.mu.Lock()
-				br.mu.caughtUp = false
+				br.testPendingEventToSend.Add(1)
 			case <-ctx.Done():
 				br.mu.Lock()
 				alloc.Release(ctx)
@@ -216,7 +217,6 @@ func (br *bufferedRegistration) outputLoop(ctx context.Context) error {
 		br.mu.Lock()
 		if len(br.buf) == 0 {
 			overflowed = br.mu.overflowed
-			br.mu.caughtUp = true
 		}
 		if firstIteration {
 			wasOverflowedOnFirstIteration = br.mu.overflowed
@@ -235,13 +235,13 @@ func (br *bufferedRegistration) outputLoop(ctx context.Context) error {
 
 		select {
 		case nextEvent := <-br.buf:
-
 			if wasOverflowedOnFirstIteration && !oneCheckpointWithTimestampSent {
 				isCheckpointEvent := nextEvent.event != nil && nextEvent.event.Checkpoint != nil
 				oneCheckpointWithTimestampSent = isCheckpointEvent && !nextEvent.event.Checkpoint.ResolvedTS.IsEmpty()
 			}
 
 			err := br.stream.SendUnbuffered(nextEvent.event)
+			br.testPendingEventToSend.Add(-1)
 			nextEvent.alloc.Release(ctx)
 			putPooledSharedEvent(nextEvent)
 			if err != nil {
@@ -308,7 +308,9 @@ func (br *bufferedRegistration) maybeRunCatchUpScan(ctx context.Context) error {
 	return catchUpIter.CatchUpScan(ctx, br.stream.SendUnbuffered, br.withDiff, br.withFiltering, br.withOmitRemote)
 }
 
-// Wait for this registration to completely process its internal buffer.
+// Wait for this registration to completely process its internal
+// buffer. This is only used when a test sends a sync event to the
+// rangefeed processor with testRegCatchupSpan set.
 func (br *bufferedRegistration) waitForCaughtUp(ctx context.Context) error {
 	opts := retry.Options{
 		InitialBackoff: 5 * time.Millisecond,
@@ -317,9 +319,7 @@ func (br *bufferedRegistration) waitForCaughtUp(ctx context.Context) error {
 		MaxRetries:     50,
 	}
 	for re := retry.StartWithCtx(ctx, opts); re.Next(); {
-		br.mu.Lock()
-		caughtUp := len(br.buf) == 0 && br.mu.caughtUp
-		br.mu.Unlock()
+		caughtUp := len(br.buf) == 0 && br.testPendingEventToSend.Load() == 0
 		if caughtUp {
 			return nil
 		}

